core/src/snapshot_packager_service/snapshot_gossip_manager.rs

```diff
@@ -26,120 +24,153 @@ impl SnapshotGossipManager {
     pub fn new(
         cluster_info: Arc<ClusterInfo>,
         max_full_snapshot_hashes: usize,
-        max_incremental_snapshot_hashes: usize,
+        _max_incremental_snapshot_hashes: usize,
     ) -> Self {
         SnapshotGossipManager {
             cluster_info,
+            latest_snapshot_hashes: None,
             max_full_snapshot_hashes,
-            max_incremental_snapshot_hashes,
-            full_snapshot_hashes: FullSnapshotHashes {
-                hashes: Vec::default(),
-            },
-            incremental_snapshot_hashes: IncrementalSnapshotHashes {
-                base: (Slot::default(), SnapshotHash(Hash::default())),
+            legacy_full_snapshot_hashes: FullSnapshotHashes {
                 hashes: Vec::default(),
             },
         }
     }
 
-    /// If there were starting snapshot hashes, add those to their respective vectors, then push
-    /// those vectors to the cluster via CRDS.
+    /// Push any starting snapshot hashes to the cluster via CRDS
     pub fn push_starting_snapshot_hashes(
         &mut self,
         starting_snapshot_hashes: Option<StartingSnapshotHashes>,
     ) {
-        if let Some(starting_snapshot_hashes) = starting_snapshot_hashes {
-            let starting_full_snapshot_hash = starting_snapshot_hashes.full;
-            self.push_full_snapshot_hash(starting_full_snapshot_hash);
+        let Some(starting_snapshot_hashes) = starting_snapshot_hashes else {
+            return;
+        };
 
-            if let Some(starting_incremental_snapshot_hash) = starting_snapshot_hashes.incremental {
-                self.push_incremental_snapshot_hash(starting_incremental_snapshot_hash);
-            };
+        self.update_latest_full_snapshot_hash(starting_snapshot_hashes.full.hash);
+        if let Some(starting_incremental_snapshot_hash) = starting_snapshot_hashes.incremental {
+            self.update_latest_incremental_snapshot_hash(
+                starting_incremental_snapshot_hash.hash,
+                starting_incremental_snapshot_hash.base.0,
+            );
         }
+        self.push_latest_snapshot_hashes_to_cluster();
+
+        // Handle legacy snapshot hashes here too
+        // Once LegacySnapshotHashes are removed from CRDS, also remove them here
+        self.push_legacy_full_snapshot_hash(starting_snapshot_hashes.full);
     }
 
-    /// Add `snapshot_hash` to its respective vector of hashes, then push that vector to the
-    /// cluster via CRDS.
+    /// Push new snapshot hash to the cluster via CRDS
     pub fn push_snapshot_hash(
```

Comment:
Would be nice to have a single public facing API here that could handle either: full, incremental, or both (starting). Outside the scope of this change